sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -539,3 +539,76 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
 
   override def prettyName: String = "str_to_map"
 }
+
+/**
+ * Adds/replaces field in struct by name.
+ */
+case class WithFields(
+    structExpr: Expression,
+    names: Seq[String],
+    valExprs: Seq[Expression]) extends Unevaluable {
+
+  assert(names.length == valExprs.length)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (!structExpr.dataType.isInstanceOf[StructType]) {
+      TypeCheckResult.TypeCheckFailure(
+        "struct argument should be struct type, got: " + structExpr.dataType.catalogString)
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override def children: Seq[Expression] = structExpr +: valExprs
+
+  private lazy val addOrReplaceExprs = names.zip(valExprs)
+
+  override def dataType: StructType = {
+    val existingStructFields: Seq[(String, StructField)] = structExpr.dataType
+      .asInstanceOf[StructType].map(f => (f.name, f))
+    val addOrReplaceStructFields: Seq[(String, StructField)] = addOrReplaceExprs.map {
+      case (name, expr) => (name, StructField(name, expr.dataType, expr.nullable))
+    }
+    StructType(addOrReplace(existingStructFields, addOrReplaceStructFields).map(_._2))
+  }
+
+  override def foldable: Boolean = structExpr.foldable && valExprs.forall(_.foldable)
+
+  override def nullable: Boolean = structExpr.nullable
+
+  override def prettyName: String = "with_fields"
+
+  lazy val evalExpr: Expression = {
+    val existingExprs = structExpr.dataType.asInstanceOf[StructType].fieldNames.zipWithIndex.map {
+      case (name, i) => (name, GetStructField(structExpr, i, Some(name)))
```

Comment:
nit: We can just write `GetStructField(structExpr, i)`

Reply:
Another idea: We can use `GetStructField(KnownNotNull(structExpr), i)`, because we only evaluate it if `IsNull(structExpr)` is false.

Then we can solve https://github.com/apache/spark/pull/27066/files#r448598639

Reply:
nice! made both changes.